components/raftstore/src/store/worker/split_controller.rs

```diff
@@ -48,56 +46,57 @@ impl Sample {
 {
-    let mut pre_sum = vec![];
     let mut sum = 0;
-    for item in iter {
+    iter.map(|item| {
         sum += read(item);
-        pre_sum.push(sum);
-    }
-    pre_sum
+        sum
+    })
+    .collect()
 }
 
-// It will return sample_num numbers by sample from lists.
-// The list in the lists has the length of N1, N2, N3 ... Np ... NP in turn.
-// Their prefix sum is pre_sum and we can get mut list from lists by get_mut.
-// Take a random number d from [1, N]. If d < N1, select a data in the first list with an equal probability without replacement;
-// If N1 <= d <(N1 + N2), then select a data in the second list with equal probability without replacement;
-// and so on, repeat m times, and finally select sample_num pieces of data from lists.
+// This function uses the distributed/parallel reservoir sampling algorithm.
+// It will sample min(sample_num, all_key_ranges_num) key ranges from multiple key_ranges_providers with the same possibility.
+// NOTICE: prefix_sum should be calculated from the lists, so they should have the same length too.
 fn sample<F, T>(
     sample_num: usize,
-    pre_sum: &[usize],
-    mut lists: Vec<T>,
-    get_mut: F,
+    prefix_sum: &[usize],
+    mut key_ranges_providers: Vec<T>,
+    key_ranges_getter: F,
 ) -> Vec<KeyRange>
 where
     F: Fn(&mut T) -> &mut Vec<KeyRange>,
 {
-    let mut rng = rand::thread_rng();
-    let mut key_ranges = vec![];
-    let high_bound = *pre_sum.last().unwrap();
-    for _ in 0..MAX_RETRY_TIME {
-        let d = rng.gen_range(0..high_bound + 1) as usize;
-        let i = match pre_sum.binary_search(&d) {
-            Ok(i) => i,
-            Err(i) => i,
-        };
-        if i < lists.len() {
-            let list = get_mut(&mut lists[i]);
-            if !list.is_empty() {
-                let j = rng.gen_range(0..list.len()) as usize;
-                key_ranges.push(list.remove(j)); // Sampling without replacement
-            }
-        }
-        if key_ranges.len() >= std::cmp::min(sample_num, high_bound) {
-            break;
-        }
+    let mut sampled_key_ranges = vec![];
+    if prefix_sum.len() != key_ranges_providers.len() {
+        error!("failed to sample, prefix_sum length is not equal to key_ranges_providers length";
+            "prefix_sum_len" => prefix_sum.len(),
+            "key_ranges_providers_len" => key_ranges_providers.len(),
+        );
+        return sampled_key_ranges;
     }
-    key_ranges
+
+    let mut rng = rand::thread_rng();
+    // The last sum of the QPS is the number of all the key ranges.
+    let all_key_ranges_num = *prefix_sum.last().unwrap();
+    let sample_num = std::cmp::min(sample_num, all_key_ranges_num);
+    while sampled_key_ranges.len() < sample_num {
+        let i = prefix_sum
+            .binary_search(&rng.gen_range(0..=all_key_ranges_num))
+            .unwrap_or_else(|i| i);
+        let key_ranges = key_ranges_getter(&mut key_ranges_providers[i]);
+        if !key_ranges.is_empty() {
```

Comment:
should we do something to control the number of loops ? Was it possible that in some case we meet lots of `key_ranges.is_empty`

Reply:
It will not last too long due to the possibility.